# All preprocessing scripts for the methylation file (Removing NAN, and near-zero values columns excluding the lower variance CpG sites ) 
```{r}
library(data.table)
library(caret)
setDTthreads(11)
data <- fread("~/master_thesis_data/processed_data.csv",data.table=FALSE)
cat("orginal num_data size: ", dim(data))

rownames(data) <- data[,1]
# get the labels 
for(i in 1:nrow(data)){
  if (grepl("R1", data[i,1], fixed = TRUE)){
    data[i,1] <- 'R'
  }
  else if (grepl("R2", data[i,1], fixed = TRUE)){
    data[i,1] <- 'R'
  }
  else if (grepl("R3", data[i,1], fixed = TRUE)){
    data[i,1] <- 'R'
  }
  else if (grepl("D", data[i,1], fixed = TRUE)){
    data[i,1] <- 'D'
  }
}
# data[,1]
num_data <- data[, -c(1)]
num_data <- na.omit(t(num_data))
num_data <- t(num_data)
cat(" num_data size: ", dim(num_data))

# nzv_col <- nearZeroVar(num_data,allowParallel = TRUE)
# data_nzv <-num_data[,-c(nzv_col)]
# cat("nearZeroVar: ", dim(data_nzv))


# df <- data[,!colnames(data) %in% c('TargetID')]
saveRDS(num_data, "~/rmcfs_on_q08/filtered_data_only.rds")
cat("quantile prob = 0.8")
col_sd <- apply(num_data, 2, sd)
sd_limit <- quantile(col_sd,prob=0.8)

indeces <- which(col_sd >= sd_limit)
filtered_data <- num_data[,c(indeces)]
filtered_data <- cbind.data.frame(filtered_data, decision = data$TargetID)
saveRDS(filtered_data, "~/rmcfs_on_q08/filtered_data_q08.rds")
```

# All ICluster Bayes scripts
## Applying Icluster_on_RMCFS
```{r}
library(iClusterPlus)
data <- readRDS("~/rmcfs_on_q08/rmcfs_met_570.rds")
met_data <- as.matrix(data)
met_data <- met_data[order(rownames(met_data)),]

gx <- readRDS("~/rmcfs_on_q08/rmcfs_gx_417.rds")
gx <- as.matrix(gx)
gx <- gx[order(rownames(gx)),]

for (i in seq(1,63,1)){
  if (rownames(gx)[i]==rownames(met_data)[i]){
    cat(rownames(gx)[i],rownames(met_data)[i],"\n")
  }
}
all(rownames(met_data)==rownames(gx))

#---- iClusterBayes ------------------------------------
set.seed(123)
date()
no_cores <- detectCores()
no_cores
cat("start running at for k: ", "at", date())
bayfit = tune.iClusterBayes(cpus=12,
                            dt1= met_data,
                            dt2=gx,
                            type=c("gaussian","gaussian"),
                            K=1:6,
                            n.burnin=30000,
                            n.draw=18000,
                            prior.gamma=c(0.5,0.5),
                            sdev=0.05,
                            thin=3)
cat("end running at for k: at", date())
save.image(file="gbmBayfit_rmcfs_data_met_570_gx_417_.RData")
```
## icluster bayes analysis
```{r}
library(iClusterPlus)
library(GenomicRanges)
library(gplots)
library(lattice)
library("IlluminaHumanMethylationEPICanno.ilm10b2.hg19")
load("~/gbmBayfit_rmcfs_data_met_570_gx_417_.RData")
met <- readRDS("~/rmcfs_on_q08/rmcfs_met_570.rds")
met <- as.matrix(met)
met <- met[order(rownames(met)),]

gx <- readRDS("~/rmcfs_on_q08/rmcfs_gx_417.rds")
gx <- as.matrix(gx)
gx <- gx[order(rownames(gx)),]

all(rownames(met)==rownames(gx))


allBIC = NULL
devratio = NULL
nK = length(bayfit$fit)
for(i in 1:nK){
  allBIC = c(allBIC,bayfit$fit[[i]]$BIC)
  devratio = c(devratio,bayfit$fit[[i]]$dev.ratio)
}
# plot BIC and Deviance ratio

par(mar=c(4.0,4.0,0.5,0.5),mfrow=c(1,2))
plot(1:nK, allBIC,type="b",xlab="k",ylab="BIC",pch=c(1,1,19,1,1,1))
plot(1:nK,devratio,type="b",xlab="k",ylab="Deviance ratio",pch=c(1,1,19,1,1,1))

par(mfrow=c(2,1))
k=3
plot(bayfit$fit[[k]]$beta.pp[[1]],xlab="Genomic region",ylab="Posterior probability",main="Methylation")
plot(bayfit$fit[[k]]$beta.pp[[2]],,xlab="Genes",ylab="Posterior probability",main="Expression")

col.scheme = alist()
col.scheme[[1]] = colorpanel(19,"blue","white","red")
col.scheme[[2]] = colorpanel(19,"blue","white","red")

chr=unlist(strsplit(colnames(met),"\\."))
chr=chr[seq(1,length(chr),by=10)]
chr=gsub("cg","",chr)
chr=as.numeric(chr)

plotHMBayes(fit=bayfit$fit[[k]],datasets=list(met,gx),type=c("gaussian","gaussian"), col.scheme = col.scheme,threshold=c(0.5,0.5),row.order=c(T,T),chr=chr,plot.chr=c(F,F),sparse=c(F,T),cap=c(F,F))

best.cluster.Bayes = bayfit$fit[[k]]$clusters
# best.fit=bayfit$fit[[which.min(allBIC[k])]]
best.fit=bayfit[["fit"]][[k]]
gx <- readRDS("~/rmcfs_on_q08/rmcfs_gx_417.rds")
features = alist()
features[[1]] = colnames(met)
features[[2]] = colnames(gx)

sigfeatures=alist()
for(i in 1:2){
  rowsum=apply(abs(best.fit$beta[[i]]),1, sum)
  upper=quantile(rowsum,prob=0.75)
  sigfeatures[[i]]=(features[[i]])[which(rowsum>upper)]
  }
names(sigfeatures)=c("Methylation","Gene expression")

#print a few examples of selected features
head(sigfeatures[[1]])

## mapping significat features to closest GENE 
annotation.table = getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b2.hg19)
mapp.cpg <- function(cpg, anno.table){
  
  cpg.gene <- data.frame()
  for(i in 1:length(cpg)){
    row <- cbind.data.frame(cpg[i],
                            anno.table[which(anno.table$Name==cpg[i]),]$UCSC_RefGene_Name,
                            anno.table[which(anno.table$Name==cpg[i]),]$chr,
                            anno.table[which(anno.table$Name==cpg[i]),]$pos,
                            anno.table[which(anno.table$Name==cpg[i]),]$Islands_Name)
    cpg.gene <- rbind.data.frame(cpg.gene, row)
  }
  colnames(cpg.gene) <- c('cpg.site', 'gene.name', 'chr', 'pos', 'Islands_Name')
  
  return(cpg.gene)
}

cpg2gene <- mapp.cpg(sigfeatures[[1]], annotation.table)

## Reshape the data to fit with the plotting functions 
dd <- matrix(0, nrow = 2, ncol = k+1)

for(i in seq(1,63,1)){
  if (grepl("R", rownames(met)[i], fixed = TRUE)){
    dd[1,bayfit[["fit"]][[k]][["clusters"]][i]] <-dd[1,bayfit[["fit"]][[k]][["clusters"]][i]] + 1
    
  }
  else if (grepl("D", rownames(met)[i], fixed = TRUE)){
   dd[2,bayfit[["fit"]][[k]][["clusters"]][i]] <-dd[2,bayfit[["fit"]][[k]][["clusters"]][i]] + 1
    
  }
  
  cat(rownames(met)[i], "cluster", bayfit[["fit"]][[k]][["clusters"]][i],"\n")
}

rownames(dd) <- c("Relapce", "Diagnose")
colnames(dd) <- c(1:(k+1))


beta.pp <- best.fit[["beta.pp"]][[1]]
top_143_met_beta.pp <- order(beta.pp, decreasing=FALSE)
top_143_met.pp <- features[[1]][top_143_met_beta.pp]
# heatmap(met_data[,features[[1]][top_134_beta.pp]],Rowv=NA, Colv = NA, cexCol =1)
for (i in seq(1,143,1)){
  if (!colnames(met)[top_143_met_beta.pp[i]]  %in% sigfeatures[["Methylation"]]){
    cat(colnames(met)[top_143_met_beta.pp[i]],"\n")
  }
}

gx <- readRDS("~/rmcfs_on_q08/rmcfs_gx_417.rds")
gx <- as.matrix(gx)
gx <- gx[order(rownames(gx)),]
beta.pp <- best.fit[["beta.pp"]][[2]]
top_103_gene_beta.pp <- order(beta.pp, decreasing=FALSE)
top_103_gene.pp <- features[[2]][top_103_gene_beta.pp]




library(pheatmap)
library(RColorBrewer)
met_data <- cbind.data.frame(met_data, clusters = bayfit$fit[[k]]$clusters)
met_data <- met_data[order(met_data$clusters),]

sum(met_data$clusters == 1)


# Rows are genomic features and columnsare samples
plotHMBayes(fit=bayfit$fit[[k]],datasets=list(met[,sigfeatures[["Methylation"]]]),type=c("gaussian","gaussian"), col.scheme = col.scheme,threshold=c(0.5,0.5),row.order=c(T,T),chr=chr,plot.chr=c(F,F),sparse=c(F,F),cap=c(F,F))

## Normalize the data and plotting 
normalize <- function(x) {
    return ((x - min(x)) / (max(x) - min(x)))
}
nor <- apply(gx,2,normalize)
plotHMBayes(fit=bayfit$fit[[k]],datasets=list(nor[,sigfeatures[["Gene expression"]]]),type=c("gaussian","gaussian"), col.scheme = col.scheme,threshold=c(0.5,0.5),row.order=c(T,T),chr=chr,plot.chr=c(F,F),sparse=c(F,F),cap=c(F,F))

plotHMBayes(fit=bayfit$fit[[k]],datasets=list(gx[,sigfeatures[["Gene expression"]]]),type=c("gaussian","gaussian"), col.scheme = col.scheme,threshold=c(0.5,0.5),row.order=c(T,T),chr=chr,plot.chr=c(F,F),sparse=c(F,F),cap=c(F,F))


plotHMBayes(fit=bayfit$fit[[k]],datasets=list(met[,top_143_met.pp]),type=c("gaussian","gaussian"), col.scheme = col.scheme,threshold=c(0.5,0.5),row.order=c(T,T),chr=chr,plot.chr=c(F,F),sparse=c(F,F),cap=c(F,F))
```


# Feature Boosting  
```{r}
library(data.table)
setDTthreads(11)

met_data <- readRDS("~/rmcfs_on_q08/filtered_data_q08.rds")
met_rmcfs <- readRDS("~/rmcfs_on_q08/rmcfs.rds")

data = met_data[,FALSE]
library(devtools)
library(R.ROSETTA)
library(VisuNet)
mylist <-data.frame()

for(i in seq(10,1000,10)){
  cat("range", 1,i,"\n")
  met.rosetta <- met_data[,colnames(met_data) %in% c(met_rmcfs[["RI"]][["attribute"]][1:i])]
  met.rosetta <- cbind.data.frame(met.rosetta, decision = met_data$decision)
  
  out.met <- rosetta(met.rosetta,classifier="StandardVoter",discrete=FALSE, discreteMethod="EqualFrequency",cvNum=5,reducer="Genetic",underSample =T,ruleFiltration=TRUE,ruleFiltrSupport=c(1,3) , discreteParam=3)
  mylist <- c(mylist, dim= dim(met.rosetta),accuracyMean= out.met$quality$accuracyMean)
  
}
saveRDS(mylist, "~/rmcfs_on_q08/mylist_met.rds")
```
## Plot Feature boosting  
```{r}
mat = data.frame(matrix(vector(), 0, 2,dimnames=list(c(), c("Test", "Avg"))),stringsAsFactors=F)
for (i in seq(3,300,3)) {
    print(c(mylist[[i-1]],mylist[[i]]))
    mat <- rbind(mat, c(mylist[[i-1]],mylist[[i]]))
}
colnames(mat) <- c("NumOfFeatures", "AverageAccuracy")
plot(mat)
```
